### PR TITLE
Oracle - many new things cumulated over many months

### LINKED ISSUE #5632: Fix the SQL creation scripts for FILETRANSFERSDB and TASTKS
I need to add the partitioning line for the creation scripts in order for the new dev database creations to be in sync with the production ones:

https://github.com/dmwm/CRABServer/blob/master/src/python/Databases/TaskDB/Oracle/Create.py
https://github.com/dmwm/CRABServer/blob/master/src/python/Databases/FileTransfersDB/Oracle/Create.py

### LINKED ISSUE #6472: document additional index for FILETRANSFERSDB
should add this to Create.py as well.
```


On 17/03/2021 16:00, Katarzyna Maria Dziedziniewicz-Wojcik wrote:
> Stefano,
> 
> it's:
> CREATE INDEX "CMS_ANALYSIS_REQMGR"."FILETRANSFERSDB_NEW2" ON "CMS_ANALYSIS_REQMGR"."FILETRANSFERSDB" ("TM_PUBLISH", "TM_PUBLICATION_STATE", "TM_TRANSFER_STATE")
> ________________________________________
> From: Stefano Belforte
> Sent: 16 March 2021 16:51
> To: Katarzyna Maria Dziedziniewicz-Wojcik
> Cc: Stefano Belforte
> Subject: Re: new index for CRAB
> 
> OK. Jokes aside, I see now that this index is called TM_WORKER_PUB_STATE
> which is fine, I wrote FILETRANSFERSDB_NEW2 becasue I saw that in this
> list of Indexes which is generated by Prajsh's stored procedure to
> drop partitions, I am slightly curious what FILETRANSFERSDB_NEW2 is then.
> Should we care to know ?
> 
> Status of indexes of FILETRANSFERSDB table Index Name   Index Type      Status
> TM_WORKER_PUB_STATE     NORMAL  VALID
> TM_CREATION_TIME_LOCAL_IDX      NORMAL  N/A
> ID_PK   NORMAL  VALID
> TM_WORKER_STATE NORMAL  VALID
> TM_TASKNAME_IDX NORMAL  VALID
> FILETRANSFERSDB_NEW2    NORMAL  VALID
> 
> On 16/03/2021 15:37, Katarzyna Maria Dziedziniewicz-Wojcik wrote:
>> Hi Stefano,
>>
>> That's a very educated guess! Possibly better than the actual code.
>> The code is:
>>
>>     CREATE INDEX "CMS_ANALYSIS_REQMGR"."TM_WORKER_PUB_STATE" ON "CMS_ANALYSIS_REQMGR"."FILETRANSFERSDB" ("TM_ASO_WORKER", "TM_PUBLICATION_STATE");
>>
>> If the value distribution of TM_PUBLICATION_STATE is similar to that of  TM_TRANSFER_STATE adding compress 2 would make sense, you're right.
>>
>> Cheers,
>> Kate
>>
>>
>> ________________________________________
>> From: Stefano Belforte
>> Sent: 16 March 2021 14:29
>> To: Katarzyna Maria Dziedziniewicz-Wojcik
>> Cc: Stefano Belforte
>> Subject: new index for CRAB
>>
>> Hi Kate,
>> about the index which you added for CRAB database in
>> https://cern.service-now.com/nav_to.do?uri=%2Fincident.do%3Fsys_id%3D5b5da594db8260107aca890b069619da
>>
>> I'd like to add this to the code used to create the DB instance,
>> both as documentation and "one never know, we may need to create
>> a new DB at some point". I am afraid that I could not find a way
>> to tell the details from SQL.
>>
>> Our current code is [1]
>>
>>       "CREATE INDEX TM_TASKNAME_IDX ON FILETRANSFERSDB (TM_TASKNAME)"
>>       "CREATE INDEX TM_WORKER_STATE ON FILETRANSFERSDB (TM_ASO_WORKER, TM_TRANSFER_STATE) COMPRESS 2"
>>
>> can you say which line should be added to reproduce what you have done ?
>> I could guess a
>> "CREATE INDEX FILETRANSFERSDB_NEW2 ON FILETRANSFERSDB (TM_ASO_WORKER, TM_PUBLICATION_STATE) COMPRESS 2"
>>
>> but it is the kind of thing where guess work is not good !
>>
>> Thanks a lot
>> Stefano
>>
>> [1]
>> https://github.com/dmwm/CRABServer/blob/dd8656160ddd453e2efd190a4497f1cf6e16a09f/src/python/Databases/FileTransfersDB/Oracle/Create.py#L26-L27
>>
> 
```

### LINKED ISSUE #7154: Oracle DB - setup all tables for partitioning
setup all tables preprod and dev DB instances for partitioning. Now only preprod FILETRANSFERDB is correct

### LINKED ISSUE #7458: Review instructions for using private Oracle DB for developers
make sure they are up to date

### PR DESCRIPTION
**Pull request by mapellidario**
fixes #7154 
fixes #6472 
fixes #5632
fixes #7458 

### description

This PR adds some things related to the oracle DB management:

- [x] Adjust Create.py so that they create partitioned tables, tested
- [x] Adjust Create.py so that they create proper indexes, same as production, tested
- [x] added partitions.sql that contains all the procedures, functions and scheduled jobs for dropping old partitions. tested.

Together with all the information at [1] this should be enough to take care of the DB for the years to come :)

This PR supersedes all the information that is contained in obsolete and archived repositories in gitlab [2].

---

[1]

- https://cmscrab.docs.cern.ch/technical/crab-rest/database-admin.html
- https://cmscrab.docs.cern.ch/technical/crab-rest/database.html

[2] https://gitlab.cern.ch/groups/crab3/-/archived

### CODE REVIEW

File: etc/updateOracle.sql [hunk 1]
```diff
@@ -139,3 +139,22 @@ ALTER TABLE tasks ADD tm_publish_rule VARCHAR(255);
 
 --Add Rucio ASO's json kv for store container names and its rules.
 ALTER TABLE tasks ADD tm_multipub_rule CLOB;
+
+-- 2024-03 oracle cleanup
```
Comment (belforte) [inline etc/updateOracle.sql:143]: I am not sure that it is worth to keep here (ever if as comments, IIUC) actions that you did just once to remove temporary but left-over tables. Maybe simply a note about this process in the doc ? (I have not read doc yet, maybe it is there already)

File: etc/oracle-admin/procedures.sql [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    sql_query := substr(sql_query, 0, length(sql_query)-1) || ' )';
+    dbms_output.put_line('partition_key: '||partition_key);
+    dbms_output.put_line('index_name: '||index_name);
+    dbms_output.put_line('sql_query: '||sql_query);
+    execute immediate sql_query;
+    dbms_output.put_line('===========================================================================');
+end;
+
+END BUILD_LOCAL_INDEX;
+/
```
Comment (belforte) [inline etc/oracle-admin/procedures.sql:34]: forgive my ignorance. I this lonely slash ('/') OK ? is a terminator for line 5 ?

File: etc/oracle-admin/procedures.sql [hunk 3]
```diff
@@ -0,0 +1,34 @@
+/*
+Some procedures that dario found on production DB, which may be useful in the future
```
Comment (belforte) [inline etc/oracle-admin/procedures.sql:2]: what about the montlhy partition dropping ? Aren't those stored procedures as well ?

File: etc/oracle-admin/partitions.sql [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+	The following PL/SQL code will be used to create 
+	an automatic process to remove 90 days older partitions 
+	in FILETRANSFERSDB and FILEMETADATA tables in production database.
+
+Database: 
+=========
+    These code is applied to the following databases:
+
+	- production:    cms_analysis_reqmgr@cmsr
+    - preproduction: cmsweb_analysis_preprod@int2r
```
Comment (belforte) [inline etc/oracle-admin/partitions.sql:13]: didnt' you add `dev` as well ?

File: etc/oracle-admin/partitions.sql [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+
+Dependency/Requirements:
+========================
+	1. We must ensure that we have proper permission to create
```
Comment (belforte) [inline etc/oracle-admin/partitions.sql:17]: care to say how we check ? WHo is "we" ? the schema owner ?

File: etc/oracle-admin/partitions.sql [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+
+
+/*=========================================================================================================================*/
+CREATE OR REPLACE PROCEDURE SEND_EMAIL
+( data  IN varchar2 )
+IS
+	username varchar2(100);
+	dbname   varchar2(100);
+	emails   varchar(200) := 'cms-service-crab-operators@cern.ch';
+	  /*emails   varchar(200) := 'alccs.prajesh.sharma@gmail.com,prajesh.sharma@cern.ch,Stefano.Belforte@cern.ch,daina.dirmaite@cern.ch';*/
```
Comment (belforte) [inline etc/oracle-admin/partitions.sql:93]: do we need to keep those lines with old emails ?

File: etc/oracle-admin/procedures.sql (hunk 3 above)
Comment (belforte) [reply etc/oracle-admin/procedures.sql:2]: OK. those are in the "partitons.sql"

**Review by belforte — APPROVED** [review]

**Comment by belforte** [comment]
@mapellidario I haven't reviewed the documentation yet, but you can still merge this

**Comment by belforte** [comment]
about documentation at the top of https://cmscrab.docs.cern.ch/technical/crab-rest/database.html#crab-database-developer
I do not understand
```
We use Oracle Database to store the task state. With the current policy: we store data for only 30 days.

It is possible to ask DBA to roll back our schemas to any point in time within the last six months. But it will spawn a new schema instead of reverting to the current one.
```
we do not loose data from DB after 30 days. And last line sort of makes no sense. "rolling back" is not "reverting to the current one"

**Comment by belforte** [comment]
https://cmscrab.docs.cern.ch/technical/crab-rest/database-faqs.html#database-faqs

what is " kill it through WebUI." ?

File: etc/updateOracle.sql (hunk 1 above)
Comment (mapellidario) [reply etc/updateOracle.sql:143]: oook! :)

File: etc/oracle-admin/procedures.sql (hunk 2 above)
Comment (mapellidario) [reply etc/oracle-admin/procedures.sql:34]: I added a comment at the beginning of every `.sql` file that points to https://stackoverflow.com/a/10207695

File: etc/oracle-admin/partitions.sql (hunk 4 above)
Comment (mapellidario) [reply etc/oracle-admin/partitions.sql:13]: ops, i forgot that i had this list as well. yes, updated, thanks!

File: etc/oracle-admin/partitions.sql (hunk 5 above)
Comment (mapellidario) [reply etc/oracle-admin/partitions.sql:17]: yeah, bad wording, i improved it

File: etc/oracle-admin/partitions.sql (hunk 6 above)
Comment (mapellidario) [reply etc/oracle-admin/partitions.sql:93]: nope

**Comment by mapellidario** [comment]
> https://cmscrab.docs.cern.ch/technical/crab-rest/database-faqs.html#database-faqs
> 
> what is " kill it through WebUI." ?

I did not write that page. However, I think it means that from the "sessionmanager web ui" that is mentioned in the sentence before. I updated that sentence and should be a bit more clear now

**Pull request merged**